### ISSUE TITLE
Coevolution in UT2004

### ISSUE THREAD

**Issue by schrum2** [schrum2/MM-NEAT#193]
Once #192 is finished you can start considering coevolution in UT2004. We will need to discuss this a lot. You will actually have to start by refactoring UT2004Task so that it still works the way we have been using it, except that whatever method ends up launching the server and doing the evaluation needs to return the whole GameDataCollector[] collectors array instead of just the first index. In the new version of the code, the oneEval method will call this method, and then return the first index.

This version of the refactored code should be tested, committed and pushed before actually moving on to coevolution. However, once we know we haven't broken anything, we can try making the game work with coevolution too. This will involve taking a portion of the population (perhaps 5 at a time ... this will probably be a command line parameter) and running them on a server against each other.

We might also want a few native and/or HunterBots around, but we can worry about those details later.

**Comment by schrum2** [comment]
With all of the refactoring I've done, you can now make some new classes in the edu.southwestern.tasks.ut2004 package. Specifically, make a class UT2004MultipleEvolvedBotsDeathMatchTask and have it extend SinglePopulationCoevolutionTask. Then fill in the necessary methods to co-evolve bots. Most of the hard work can be accomplished by simply calling static methods in the refactored UT2004Task.

For example, the evaluateGroup method should just call the UT2004Task.evaluateMultipleGenotypes method, and then put the pair results into Score instances. Look at line 187 of NoisyLonerTask to see how this is done.

You will also need to define the groupSize method, but this should probably just return the value assigned to some command line Parameter that you will need to make.

**Comment by friedmaa** [comment]
Did not get to this.

**Comment by schrum2** [comment]
This would be a more efficient utilization of UT2004 server evaluation time, but I'm not sure the results would actually be better. Although it would be nice to do at a later date, it is on hold for now.

**Comment by schrum2** [comment]
No one is going to do this.
